Add PHPstan and move code quality tools to subfolder

bezin opened the pull request:
Adds basic PHPstan configuration to the skeleton bundle

leofeyer commented:
If we keep adding CI related stuff, we should probably promote using `bamarni/composer-bin-plugin` and move PHPStan and ECS into a `tools` subfolder (as we do in `contao/contao`).

bezin commented:
I moved ECS and PHPstan to the tools subfolder and use the `composer-bin-plugin`. Like this setup a lot.

bytehead commented:
What about the CI tasks for PHPStan and CS?

bezin commented:
> What about the CI tasks for PHPStan and CS?

My main goal was to have these tools pre-configured at hand on the CLI. I wonder if we should add GitHub specific stuff? I mean, sure, everything else is already opinionated, too, but then, I use Bitbucket to host private bundles I develop for clients 🤔

bytehead commented:
I was more thinking about the bundle itself to have PHPStan and CS checks in the future. :)

bezin commented:
Aye, that makes sense 👍

bezin commented:
Would be nice if somebody could check if the workflow file is alright. I only copied it from the monorepo and adjusted it to my best knowledge while having zero experience with Github workflows😉

ausi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,62 @@
+name: CI
+
+on:
+    pull_request: ~
+    push:
+        branches:
+            - main
+        tags:
+            - '*'
+    schedule:
+        - cron: 0 13 * * MON,THU
+
```
```suggestion
on:
    pull_request: ~
    push: ~
```

`schedule` is very specific IMO, triggering on every push and on every pull request seems like a good default to me.

bezin replied:
See [51bc1b9](https://github.com/contao/skeleton-bundle/pull/11/commits/51bc1b9f6fc4f13ca7a865caec4ac807716668bd)

ausi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,62 @@
+        - cron: 0 13 * * MON,THU
+
+jobs:
+    ecs:
```
We should also add a job that runs the unit tests for every supported PHP version I think.

leofeyer replied:
ECS should be run with the lowest supported PHP version only, so there will be no incompatible changes.

bezin replied:
Ok, I think these are two different things: 

* ECS: Will update ECS to run with PHP 7.4. What about PHPStan? Newest PHP version or lowest supported?
* Unit tests: I copied them from the monorepo as well, but thought it was over the top. On the other hand: What's the harm? Will add them, too.

ausi replied:
For phpstan usually the newest I think :)

leofeyer commented on .gitignore:
```diff
@@ -5,3 +5,7 @@
 /phpunit.xml
+
+# Tools
+/tools/*/composer.lock
+/tools/*/vendor
```
```suggestion
/tools/*/vendor
```

The composer.lock file should not be excluded, especially when the bundle is reusable.

bezin replied:
See [09a11ec](https://github.com/contao/skeleton-bundle/pull/11/commits/09a11eccc04f15ca972c207eb94c50c0a8d0e406)

ausi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,107 @@
+    ecs:
+        name: ECS
+        runs-on: ubuntu-latest
+        if: github.event_name != 'push'
```
I’d remove this check (is specific to contao/contao IMO)

ausi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,107 @@
+    tests:
+        name: PHP ${{ matrix.php }}
+        runs-on: ubuntu-latest
+        if: github.event_name != 'push'
```
Same here :)

ausi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,107 @@
+    nightly:
+        name: PHP 8.2
+        runs-on: ubuntu-latest
+        if: github.event_name != 'push'
```
:)

ausi reviewed (approved):
❤️

leofeyer commented:
Somehow Github actions does not work: https://github.com/contao/skeleton-bundle/pull/11/checks 🤔

bezin commented:
Maybe it needs to be merged first?

leofeyer commented:
I don‘t think so. The workflow is executed, but the parser apparently does not find any jobs.

<img width="456" alt="" src="https://user-images.githubusercontent.com/1192057/172818966-b297da18-61ce-4229-a984-73b17b1348cc.png">

leofeyer commented:
Ah, it seems that the workflow was executed in your repo. There is also an error message there: https://github.com/bezin/skeleton-bundle/actions/runs/2466969485

bezin commented:
I debugged and fixed it, but now I am confused more than before: I fixed it by skipping PhpUnitExpectationFixer as described in [this issue](https://github.com/symplify/symplify/issues/3130). But why is this not failing in the monorepo, even though we require the same version of contao/easy-coding-standards? 

Did some more research: Because in the monorepo, even though 4.13 only requires PHP 7.4, [we run ECS with PHP 8.1](https://github.com/contao/contao/blob/4.13/.github/workflows/ci.yml). Shouldn't this be running with 7.4. too?

bytehead commented on ecs.php:
```diff
@@ -6,11 +6,17 @@
 
+    if (PHP_VERSION_ID < 80000) {
+        $ecsConfig->ruleWithConfiguration(\PhpCsFixer\Fixer\ControlStructure\TrailingCommaInMultilineFixer::class, ['elements' => ['arrays'], 'after_heredoc' => true]);
+        // SKip PhpUnitExpecationFixer (see https://github.com/symplify/symplify/issues/3130)
```
```suggestion
        // Skip PhpUnitExpecationFixer (see https://github.com/symplify/symplify/issues/3130)
```

Typo 😎

bytehead reviewed (approved):
❤️

leofeyer commented:
Thank you @bezin.
